### PR TITLE
GuiAPI

### PR DESCRIPTION
**Pull request by thecatcore**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -66,6 +66,7 @@ repositories {
 	flatDir {
 		dirs "./remapped/client"
 		dirs "./remapped/server"
+		dirs "./libs"
```
Comment (halotroop2288) [inline build.gradle:69]: I don't like this idea. Can these libs be removed in favor of a custom implementation? Or at least be found on [some obscure Maven repo](https://mvnrepository.com/artifact/de.matthiasmann.twl/core/1.0.0)?

I already have concerns about using old mod code without permission (I did plan on asking for permission, but it's a little late for that.)
I don't even know the licenses of these libraries.
Apparently TWL has a [crayon license with a notice of license clause](https://github.com/MatthiasMann/twl/blob/master/src/de/matthiasmann/twl/license.txt).

File: src/main/java/guiapi/GuiModScreen.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+				GuiWidgetScreen.getInstance().minecraftInstance.options,
+				GuiWidgetScreen.getInstance().minecraftInstance.actualWidth,
+				GuiWidgetScreen.getInstance().minecraftInstance.actualHeight
+		);
+		RenderScale.scale = var5.scale;
+		var4.syncViewportSize();
+		GuiWidgetScreen.getInstance().gui.update();
+	}
+
+	public void method_130() {
```
Comment (halotroop2288) [inline src/main/java/guiapi/GuiModScreen.java]: If this overrides something, annotate it. If not, rename / remove it.

File: src/main/java/guiapi/GuiModScreen.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+import de.matthiasmann.twl.Widget;
+import de.matthiasmann.twl.renderer.lwjgl.LWJGLRenderer;
+import de.matthiasmann.twl.renderer.lwjgl.RenderScale;
+
+import net.minecraft.client.Minecraft;
+import net.minecraft.client.gui.screen.Screen;
+import net.minecraft.client.util.ScreenScaler;
+
+public class GuiModScreen extends Screen {
```
Comment (halotroop2288) [inline src/main/java/guiapi/GuiModScreen.java]: Yucky name. Probably just ModScreen? Same goes for other "Gui" prefixes.

File: src/main/java/guiapi/WidgetText.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
+package guiapi;
+
+import de.matthiasmann.twl.EditField;
+import de.matthiasmann.twl.Label;
+import de.matthiasmann.twl.model.StringModel;
+import de.matthiasmann.twl.utils.CallbackSupport;
+
+public class WidgetText extends WidgetSetting implements StringModel {
```
Comment (halotroop2288) [inline src/main/java/guiapi/WidgetText.java]: Prefix -> Suffix (and remap) for all relevant classes. (sorry)

File: src/main/resources/fabric.mod.json [hunk 5]
```diff
@@ -123,7 +127,33 @@
       "dq": "io/github/betterthanupdates/shockahpi/item/ShockAhPIToolItem",
       "ta": "io/github/betterthanupdates/shockahpi/item/ShockAhPIAxeItem",
       "au": "io/github/betterthanupdates/shockahpi/item/ShockAhPIPickaxeItem",
-      "wc": "io/github/betterthanupdates/shockahpi/item/ShockAhPIShovelItem"
+      "wc": "io/github/betterthanupdates/shockahpi/item/ShockAhPIShovelItem",
+      "GuiApiHelper": "guiapi/GuiApiHelper",
```
Comment (halotroop2288) [inline src/main/resources/fabric.mod.json:133]: Don't forget to address my earlier comments here, or you could break functionality.

File: src/main/resources/font.fnt [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    <char id="121" x="72" y="56" width="5" height="8" xoffset="1" yoffset="1" xadvance="6" page="0" chnl="0" />
+    <char id="122" x="80" y="56" width="5" height="8" xoffset="1" yoffset="1" xadvance="6" page="0" chnl="0" />
+    <char id="123" x="88" y="56" width="5" height="12" xoffset="1" yoffset="2" xadvance="8" page="0" chnl="0" />
+    <char id="124" x="96" y="56" width="1" height="12" xoffset="2" yoffset="2" xadvance="5" page="0" chnl="0" />
+    <char id="125" x="104" y="56" width="5" height="12" xoffset="2" yoffset="2" xadvance="8" page="0" chnl="0" />
+    <char id="126" x="112" y="56" width="8" height="3" xoffset="1" yoffset="7" xadvance="10" page="0" chnl="0" />
+    
+    
+  </chars>
+  <kernings count="0">
```
Comment (halotroop2288) [inline src/main/resources/font.fnt:114]: Can't comment on binary files, but you should remove the textures (and remove them from git history). They're vanilla textures, ya silly.

Except `twlGuiThemeIndentedbuttons.png`

**Review by halotroop2288 — CHANGES_REQUESTED** [review]
Rebase.

File: build.gradle (hunk 1 above)
Comment (thecatcore) [reply build.gradle:69]: The thing is we don't know at all what version of those library is used here. So I extracted the library's classes from GuiAPI jar to be sure to be using the right ones...

Comment (thecatcore) [reply build.gradle:69]: TWL's website seems to be gone too.

Comment (halotroop2288) [reply build.gradle:69]: Have you tried just using the latest version (`-SNAPSHOT`)?

File: src/main/resources/font.fnt (hunk 6 above)
Comment (thecatcore) [reply src/main/resources/font.fnt:114]: Fixed

File: src/main/java/guiapi/GuiApiHelper.java [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+			return new TextArea(model);
+		}
+	}
+
+	public static void setTextAreaText(TextArea textArea, String text) {
+		TextAreaModel model = textArea.getModel();
+
+		if (model instanceof SimpleTextAreaModel) {
+			((SimpleTextAreaModel) model).setText(text, false);
+		} else {
```
Comment (halotroop2288) [inline src/main/java/guiapi/GuiApiHelper.java]: Check type before casting. Could be overridden.

File: src/main/java/guiapi/GuiApiHelper.java [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+import de.matthiasmann.twl.textarea.SimpleTextAreaModel;
+import de.matthiasmann.twl.textarea.TextAreaModel;
+import guiapi.widget.SimpleWindowWidget;
+import guiapi.widget.SingleColumnWidget;
+
+public class GuiApiHelper {
+	public static final ModAction backModAction = new ModAction(ModScreen.class, "back");
+	public static final ModAction clickModAction = new ModAction(ModScreen.class, "clicksound");
+	private ArrayList<AbstractMap.SimpleEntry<String, ModAction>> buttonInfo_;
+	private String displayText_;
```
Comment (halotroop2288) [inline src/main/java/guiapi/GuiApiHelper.java]: underscores BAD

File: src/main/java/guiapi/ModAction.java [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+		try {
+			Method meth = this.GetMethodRecursively(this.objectRef, this.methodName);
+			return meth.invoke(this.objectRef instanceof Class ? null : this.objectRef, args);
+		} catch (Exception var3) {
+			throw new Exception("error calling callback '" + this.nameRef + "'.", var3);
+		}
+	}
+
+	protected Method GetMethodRecursively(Object o, String method) {
```
Comment (halotroop2288) [inline src/main/java/guiapi/ModAction.java]: Move this to Apron's `ReflectionUtils` class.

File: src/main/java/guiapi/ModScreen.java [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+				WidgetScreen.getInstance().minecraftInstance.actualWidth,
+				WidgetScreen.getInstance().minecraftInstance.actualHeight
+		);
+		RenderScale.scale = var5.scale;
+		var4.syncViewportSize();
+		WidgetScreen.getInstance().gui.update();
+	}
+
+	@Override
+	public void method_130() {
```
Comment (halotroop2288) [inline src/main/java/guiapi/ModScreen.java:82]: wtf is this tho

File: src/main/java/guiapi/WidgetScreen.java [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+import de.matthiasmann.twl.GUI;
+import de.matthiasmann.twl.Widget;
+import de.matthiasmann.twl.input.lwjgl.LWJGLInput;
+import de.matthiasmann.twl.renderer.lwjgl.LWJGLRenderer;
+import de.matthiasmann.twl.theme.ThemeManager;
+
+import net.minecraft.client.Minecraft;
+import net.minecraft.client.util.ScreenScaler;
+
+public class WidgetScreen extends Widget {
```
Comment (halotroop2288) [inline src/main/java/guiapi/WidgetScreen.java]: ScreenWidget

File: src/main/java/guiapi/WidgetScreen.java [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+	public static WidgetScreen instance;
+	public static int screenheight;
+	public static int screenwidth;
+	public Widget currentWidget = null;
+	public GUI gui = null;
+	public Minecraft minecraftInstance;
```
Comment (halotroop2288) [inline src/main/java/guiapi/WidgetScreen.java]: client

File: src/main/java/io/github/betterthanupdates/guiapi/mixin/client/OptionsScreenMixin.java [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+import net.minecraft.client.gui.screen.Screen;
+import net.minecraft.client.gui.screen.menu.OptionsScreen;
+import net.minecraft.client.gui.widget.ButtonWidget;
+
+@Mixin(OptionsScreen.class)
+public class OptionsScreenMixin extends Screen {
+	@Inject(method = "init", at = @At("RETURN"))
+	private void guiapi$init(CallbackInfo ci) {
+		this.buttons.add(new ButtonWidget(300, this.width / 2 - 100, this.height / 6 + 192, "Global Mod Settings"));
```
Comment (halotroop2288) [inline src/main/java/io/github/betterthanupdates/guiapi/mixin/client/OptionsScreenMixin.java:21]: TODO: an API for adding buttons to the menu.

TODO: This string needs to be localized, eventually.

File: src/main/resources/font.fnt (hunk 6 above)
Comment (halotroop2288) [reply src/main/resources/font.fnt:114]: gui.png still exists.

Comment (halotroop2288) [reply src/main/resources/font.fnt:114]: font/default.png also

**Review by halotroop2288 — CHANGES_REQUESTED** [review]

File: src/main/java/guiapi/ModScreen.java (hunk 10 above)
Comment (thecatcore) [reply src/main/java/guiapi/ModScreen.java:82]: idk

File: src/main/java/guiapi/WidgetScreen.java (hunk 12 above)
Comment (thecatcore) [reply src/main/java/guiapi/WidgetScreen.java]: public API :/

File: src/main/resources/font.fnt (hunk 6 above)
Comment (thecatcore) [reply src/main/resources/font.fnt:114]: if I remove them, the game crash because it can't find them in the same jar as the theme file...

**Pull request merged**